KEP-127: graduate to Beta for 1.30 (#4439)
kubernetes/enhancements: keps/sig-node/127-user-namespaces/README.md

##### Beta

> - Gather and address feedback from the community

Reviewer: Windows support was resolved several months ago. Windows maintainers commented that they don't want this feature in windows and it looked fine anyways from their point of view.

We will try to gather feedback from VM runtimes, yes

> - Gather and address feedback from the community

Reviewer: I'll try to update with the links later

> - Gather and address feedback from the community

Reviewer 2: "don't want this feature in windows"

Love it.

> - Gather and address feedback from the community

Reviewer 3: That SGTM - just please update the KEP to reflect it :)

- Be able to configure UID/GID ranges to use for pods
- Get review from VM container runtimes maintainers (not blocker, as VM runtimes should just ignore

>   the field, but nice to have)

Reviewer 4: @fidencio @littlejawa have you or anyone from kata community taken a look?

> ##### Beta

Reviewer: Added, thanks!

### Upgrade / Downgrade Strategy

Existing pods will still work as intended, as the new field is missing there.

Upgrade will not change any current behaviors.

When the new functionality wasn't yet used, downgrade will not be affected.

On downgrade, when the functionality was used, the pods created with
user namespaces that are running will continue to run with user
namespaces. Pods will need to be re-created to stop using the user
namespace.

Versions of Kubernetes that doesn't have this feature implemented will
ignore the new field `pod.spec.hostUsers`.

> 

Reviewer 3: I already commented on that in https://github.com/kubernetes/enhancements/pull/4439/#discussion_r1475820339
but it got lost.

Can you please update verskiew skew strategy to cover also kube-apiserver (that may support the new feature whereas kubelet don't have it or have FG disabled)?

Reviewer: Added now, sorry! It will be in the next push

  CRI or CNI may require updating that component before the kubelet.
-->

#### Kubelet and Kube-apiserver skew

The apiserver and kubelet feature gate enablement work fine in any combination:

1. If the apiserver has the feature gate enabled and the kubelet doesn't, then the pod will show
   that field and the kubelet will ignore it. Then, the pod is created without user namespaces.
2. If the apiserver has the feature gate disabled and the kubelet enabled, the pod won't show this
   field and therefore the kubelet won't act on a field that isn't shown. The pod is created without
   user namespaces.

The kubelet can still create pods with user namespaces if static-pods are configured with
pod.spec.hostUsers and has the feature gate enabled.

If the kube-apiserver doesn't support the feature at all (< 1.25), a pod with userns will be

> rejected.

Reviewer 3: This is not true - kube-apiserver is not rejecting the pod, but rather is silently dropping this field.

Also - this isn't only about the release <1.25, but rather also when the feature gate is disabled.

Reviewer: No, it is rejecting the pod. This is what happens when a field is known, for example:

```
$ kubectl apply -f mypod2.yaml 
Error from server (BadRequest): error when creating "mypod2.yaml": Pod in version "v1" cannot be handled as a Pod: strict decoding error: unknown field "spec.nonExistingField"
```

I added the `pod.spec.nonExistingField: true` to the pod.

Am I missing something? Or are we talking about different things?

Regarding the feature gate disabled, that is mentioned a few lines above: https://github.com/kubernetes/enhancements/pull/4439/commits/3766329491a71f33501ed417c8d34640258fffeb#diff-a9ca9fbce1538447b92f03125ca2b8474e2d875071f1172d2afa0b1e8cadeabaR684-R686

Reviewer 3: This is strange - the whole point of having the dropDisabledFields pattern is to handle and drop the field if the FG is disabled.
@liggitt - can you please shed some light on it?

Reviewer: For context, this is the apiserver as started with the local-up-cluster.sh

Reviewer 5: 1. kubectl opts into strict mode, where unknown fields are flagged / rejected. `kubectl apply ... --validate=false` shows what normal API clients (like controllers) would experience, where unknown fields are just dropped.
2. emulating the behavior where unknown fields are dropped on create when a feature is disabled is the standard approach taken.
3. *rarely*, for a security-related feature, we will reject attempts to use the new security-related field if the feature is disabled rather than dropping the field. The only one I can remember off the top of my head was the windows `hostProcess` field

Reviewer 4: do we even need this sentence? 1.25 has been out of support for a while, and we'd also not support a 1.30->1.25 skew between kubelet and KAS

Reviewer 5: > We aren't doing anything special in the apiserver for this field (pod.spec.hostUsers) on the apiserver.

Once you add the field, you need to start doing things with it in the apiserver when the feature is disabled.

Reviewer 4: Updated to specify the kube-apiserver will drop the field if it's unknown

Reviewer 4: how does it look @liggitt @wojtek-t

> If the kubelet doesn't support the feature (< 1.25), it will ignore the pod.spec.hostUsers field.

Reviewer 3: If Kubelet is on old version not knowing the field - this is fine.

I'm more interested what happens if Kubelet is new enough and understands this field (say even 1.30), but it has feature gate disabled?
Will it ignore the field completely? [this isn't expected]
Or rather, will it reject the pod because it can't fulfill it? - which is what I would expect.

Reviewer 6: We should reject the pod in that case.

Reviewer: The current proposal is that it will ignore it completely. So feature disabled and an old kubelet that doesn't even know about this field, behave the same way.

We can change it to reject the pod in that case too (it is trivial). I proposed it this way, so by disabling the feature gate you can really stop any bleeding related to user namespaces. It seems more risky in case there is some bug, IMHO.

We can change it to that if you prefer, of course. As I said, it is trivial to do that too.

Reviewer 3: I believe that we should reject (as I wrote above).

Reviewer: Perfect, changed the text to do that instead. Thanks! :)

-->

- [x] Feature gate (also fill in values in `kep.yaml`)

>   - Feature gate name: UserNamespacesPodsSupport

Reviewer 2: The gate seems to actually be named "UserNamespacesSupport"

>   - Feature gate name: UserNamespacesPodsSupport

Reviewer 4: @wojtek-t for 1 and 2: https://github.com/kubernetes/enhancements/pull/4439/files#diff-a9ca9fbce1538447b92f03125ca2b8474e2d875071f1172d2afa0b1e8cadeabaL613-L633 What are you looking for from this section that isn't covered already?

>   - Feature gate name: UserNamespacesPodsSupport

Reviewer 3: 1. Update/Downgrade is not filled in at all
2. For version skew I explicit wrote  - please describe the case what happens if feature is enabled in kube-apiserver and not in kubelet and vice-versa.

>   - Feature gate name: UserNamespacesPodsSupport

Reviewer: 1. Oh, that is one of the few section that doesn't have any comments, it is such a small text that I didn't see it! We will add it now, sorry!

###### How can a rollout or rollback fail? Can it impact already running workloads?

The rollout is just a feature flag on the kubelet and the kube-apiserver.

If one API server is upgraded while others aren't, the pod will be accepted (if the apiserver is >=
1.25). If it is scheduled to a node that the kubelet has the feature flag activated and the node
meets the requirements to use user namespaces, then the pod will be created with the namespace. If

> it is scheduled to a node that has the feature disabled, it will be scheduled without the user

Reviewer 6: scheduled or created?

Author: created

namespace.

On a rollback, pods created while the feature was active (created with user namespaces) will have to

> be restarted to be re-created without user namespaces. Just a re-creation of the pod will do the

Reviewer: Good point!

###### What specific metrics should inform a rollback?

On Kubernetes side, the kubelet should start correctly.

On the node runtime side, a pod created with pod.spec.HostUsers=false should be on RUNNING state if

> all node requirements are met.

Reviewer 3: I'm not a fan of this answer - if I'm managing 1000s of clusters, it's not something I can reasonably monitor. Can we rely on easier to monitor conditions - ideally some metrics?

Reviewer: We can add metrics to signal on the errors listed here in the KEP (in the section "What are other known failure modes?"). 

This can potentially cover a lot at the kubelet layer. But things like one of the the filesystems used by the pod doesn't support idmap mounts on the kernel running will not be possible to detect from the kubelet. The runtime will return a clear error on that case, that is shown to the user, but I don't think we can have a metric for that.

Does doing that sounds like a good plan for you?

I'm not convinced about this, though. All errors are returned to the user, so it's very visible if it fails. And having a proliferation of metrics can also harm, especially long-term. Do you think this is definitely a "yes, this metrics are useful" or do you have mixed feelings about it too?

Also, do you think this is needed for beta graduation or can be added while the feature is in beta and still disabled by default?

We will add.

> We will test for when the field pod.spec.hostUsers is set to true, false

Reviewer 3: See also the comment from the KEP teplate:
```
Additionally, for features that are introducing a new API field, unit tests that
are exercising the `switch` of feature gate itself (what happens if I disable a
feature gate after having objects written with the new field) are also critical.
You can take a look at one potential example of such test in:
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
```

that is relevant here and extend this answer.

Also - can you please add enablement/disablement tests eplicitly as beta graduation criteria?

Reviewer: Oh, great call. Added that to this section and to the beta criteria!

###### How can a rollout or rollback fail? Can it impact already running workloads?

The rollout is just a feature flag on the kubelet and the kube-apiserver.

> If one APIserver is upgraded while other's aren't and you are talking to a not upgraded the pod

Reviewer 3: I would remove this first sentence - I don't think it matters really and it doesn't bring value.

Reviewer: Okay. Sorry, I thought it would help as context, as there are lot of keps that added a field already and we all rely on existing plumbing to handle this correctly. But removed :)

will be accepted (if the apiserver is >= 1.25). If it is scheduled to a node that the kubelet has
the feature flag activated and the node meets the requirements to use user namespaces, then the
pod will be created with the namespace. If it is scheduled to a node that has the feature disabled,

> it will be created without the user namespace.

Reviewer 3: I hope it's no longer true after our discussion - the pod should fail to start in the second case, right?

Reviewer: @wojtek-t No, this is still true. If the feature is disabled in the kubelet, the kubelet completely ignores that field. The pod will only fail if the kubelet has the feature activated and the container runtime doesn't support it.

I don't think the kubelet can act on a field that should only act if the feature is enabled. We use this a lot on the error cases mentioned in detail in some other section, where you can just disable the feature gate to stop any bleeding that there might be related to this.

On a rollback, pods created while the feature was active (created with user namespaces) will have to

> be re-created without user namespaces.

Reviewer 3: ...  If those weren't recreated, they will continue to run in user namespace.

###### How can an operator determine if the feature is in use by workloads?

> Check if any pod has the pod.spec.HostUsers field set to false.

Reviewer 7: (nit)
```suggestion
Check if any pod has the `.spec.hostUsers` field set to false.
```

In the API, the field is `.spec.hostUsers`.

###### What specific metrics should inform a rollback?

On Kubernetes side, the kubelet should start correctly.

On the node runtime side, a pod created with pod.spec.hostUsers=false should be on RUNNING state if
all node requirements are met. If the CRI runtime or the handler do not support the feature, the kubelet

> returns an error.

Reviewer 3: This puts the pod into Failed state, right?

So the number of Failed pods is effectively what we can look at, right?

[Not a native metric, but something measurable]

Reviewer: I've added the following to clarify. 

---

When a pod hits this error returned by the kubelet, the status in `kubectl` is shown as
`ContainerCreating` and the pod events shows:

```
  Warning  FailedCreatePodSandBox  12s (x23 over 5m6s)  kubelet            Failed to create pod sandbox: user namespaces is not supported by the runtime
```

The following kubelet metrics are useful to check:
 - `kubelet_running_pods`: Shows the actual number of pods running
 - `kubelet_desired_pods`: The number of pods the kubelet is _trying_ to run

If these metrics are very different, it means there are desired pods that can't be set to running.
If that is the case, checking the pod events to see if they are failing for user namespaces reasons
(like the errors shown in this KEP) is advised, in which case it is recommended to rollback or
disable the feature gate.

###### How can someone using this feature know that it is working for their instance?

> Check if any pod has the pod.spec.HostUsers field set to false and is on RUNNING state on a node

Reviewer 7: ```suggestion
Check if any pod has the `.spec.ghstUsers` field set to false and is on RUNNING state on a node
```

> that meets all the requirements.

Reviewer: I can add the link to the k8s doc if that helps: https://kubernetes.io/docs/concepts/workloads/pods/user-namespaces/#before-you-begin

> 

Reviewer: @wojtek-t Why do we need that? I'm honestly not sure I understand, let me explain why:

* We are not doing enforcement of any kind based on the feature gate `UserNamespacesSupport`, so there are no security concerns for requesting this and then not being enabled.
* AFAIK we don't have that metric in the kubelet for any other namespace (nor share the PID namespace, that was added in like 1.17, quite "recently"). **Why is this namespace special?** I honestly don't see it.
* Even if we do want a metric, isn't the container runtime a better place for this metric?
* The CRI runtime doesn't really know if this was honored or not. For runc-compatible runtimes, _maybe_ we can manage to get it (using the features subcommand if idmap is supported, then it must error-out if its not honored). But there is no real feedback from the OCI runtime on what was honored or not, let alone non-OCI runtimes. 

So, I really want to push back a little here and understand why we need this and why we need this in the kubelet metrics.

> 

Reviewer 2: > Why is this namespace special? 

It's not special, but past mistakes don't justify future mistakes.

I think we have a real plumbing problem, though.  If you have a CRI which does not understand this new field, it can't report that it doesn't understand it...because it doesn't know what it doesn't know.  If we don't have a way to interrogate the runtime, how can we get this info (except to try it and look at the results)?

IOW, ideally it would be something like:

kubelet: Dear CRI, do you support userns?
runtime: No
kubelet: Dear PodStatus, this pod asked for userns but did not get it

But if there isn't a way to ask the question, it's a lot more complicated

> 

Reviewer 6: We can use RuntimeConfig CRI API for this - https://github.com/kubernetes/kubernetes/commit/0f05a92ce68ec8ae5b65aa4db746300aba5121ca

> 

Reviewer 3: Thanks @thockin for making this explicit - this is exactly my concern.
I agree that this feature isn't special, but exactly - the fact that we did something bad wrong in the past doesn't mean we can repeat mistakes.

If as a user I'm asking for X (in this case user namespace) and my pod gets to RUNNING state, I really expect it to be running in user namespace. If it's not (and it silently dropped that intent), the actual impact doesn't matter that much - it's just bad because we didn't something else than the user asked.
I would expect the pod to be rejected, and if not, at the very very least see a easy to digest for the operator signal that shows that (e.g. a metric on kubelet and a status on the pod)....

> 

Author: would you be ok with something along the lines of the following check?

```diff
diff --git a/pkg/kubelet/kuberuntime/util/util.go b/pkg/kubelet/kuberuntime/util/util.go
index af61e17567e..a9f0a33c5d5 100644
--- a/pkg/kubelet/kuberuntime/util/util.go
+++ b/pkg/kubelet/kuberuntime/util/util.go
@@ -18,6 +18,7 @@ package util
 
 import (
        v1 "k8s.io/api/core/v1"
+       internalapi "k8s.io/cri-api/pkg/apis"
        runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
        "k8s.io/klog/v2"
        kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
@@ -99,11 +100,20 @@ func PidNamespaceForPod(pod *v1.Pod) runtimeapi.NamespaceMode {
 
 // namespacesForPod returns the runtimeapi.NamespaceOption for a given pod.
 // An empty or nil pod can be used to get the namespace defaults for v1.Pod.
-func NamespacesForPod(pod *v1.Pod, runtimeHelper kubecontainer.RuntimeHelper) (*runtimeapi.NamespaceOption, error) {
+func NamespacesForPod(ctx context.Context, pod *v1.Pod, runtimeHelper kubecontainer.RuntimeHelper, runtimeService internalapi.RuntimeService) (*runtimeapi.NamespaceOption, error) {
        userNs, err := runtimeHelper.GetOrCreateUserNamespaceMappings(pod)
        if err != nil {
                return nil, err
        }
+       if userNs != nil || userNs.Mode != runtimeapi.NamespaceMode_NODE {
+               cfg, err := runtimeService.RuntimeConfig(ctx.Background())
+               if err != nil {
+                       return nil, err
+               }
+               if !runtime.Linux.UserNamespaceSupported {
+                       return nil, fmt.Errorf("The runtime doesn't support user namespaces")
+               }
+       }
```

> 

Reviewer: Note: this modulo the feature gate checking :)

> 

Author: instead of using a single bool for the RuntimeConfig, we can use the same RuntimeHandlerFeatures mechanism described for KEP-3857 (Recursive Read-Only mounts): https://github.com/kubernetes/enhancements/pull/3858/files#diff-057b8627f24bc6a0742b51b4fea113938a3440eafe255edecc834f3301c008fcR478 so we can specify the UserNamespace feature support for each separate handler

> 

Reviewer 6: Yes, that approach will work for both the cases and also handle runtime class specific features.

> 

Reviewer 3: > Just to be clear: this will only create the pod if the CRI runtime knows about userns, and return the error otherwise.

Yes - that's exactly what I would like to have.
Can you please update the PRR (and probably mention it in the core of the KEP too) to reflect that?

> 

Reviewer: Sure! Now that we agreed on a direction, we will update the KEP :)

> 

Reviewer: We updated the kep @wojtek-t , PTAL!

> 

Reviewer 3: Thank you - resolving and will add more targetted comments.

> There are step-by-step examples in the Kubernetes documentation too.

Reviewer 3: Link?

Reviewer: I can add a link to this: https://kubernetes.io/docs/tasks/configure-pod-container/user-namespaces/

I didn't do it because links here can easily become stale over time.

Reviewer: Added

###### How can a rollout or rollback fail? Can it impact already running workloads?

If one APIserver is upgraded while other's aren't and you are talking to a not
upgraded one, the pod will be accepted (if the apiserver is >= 1.25, rejected if
< 1.25).

If it is scheduled to a node where the kubelet has the feature flag activated
and the node meets the requirements to use user namespaces, then the pod will be
created with the namespace. If it is scheduled to a node that has the feature

> disabled, it will be created without the user namespace.

Reviewer 3: Same comment as for version skew.

Reviewer: Idem.

Reviewer: Fixed to that!

-->

- [ ] Events
  - Event Reason:
- [ ] API .status
  - Condition name:
  - Other field:
- [x] Other (treat as last resort)

>   - Details: check pods with pod.spec.HostUsers field set to false, and see if they are in RUNNING

Reviewer 7: ```suggestion
  - Details: check pods with `.spec.hostUsers` field set to false, and see if they are in RUNNING
```

###### How can someone using this feature know that it is working for their instance?

> If the runtime doesn't support user namespaces an error is returned by the kubelet.

Reviewer 3: So the pod will effectively transition to Failed state, right?

Reviewer: Yes, there is a FailedCreatePodSandBox event. kubectl is shown as
ContainerCreating and the pod events shows:

```
  Warning  FailedCreatePodSandBox  12s (x23 over 5m6s)  kubelet            Failed to create pod sandbox: user namespaces is not supported by the runtime
```

  - Detection: How can it be detected via metrics? Stated another way:
    how can an operator troubleshoot without logging into a master or worker node?

        See the pod events, it fails with:

    	  Warning  Failed     2s (x2 over 4s)  kubelet, 127.0.0.1  Error: failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: failed to fulfil mount request: failed to set MOUNT_ATTR_IDMAP on /var/lib/kubelet/pods/f037a704-742c-40fe-8dbf-17ed9225c4df/volumes/kubernetes.io~empty-dir/hugepage: invalid argument (maybe the source filesystem doesn't support idmap mounts on this kernel?): unknown

        Note the "maybe the source filesystem doesn't support idmap mounts on this kernel?" part.

  - Mitigations: What can be done to stop the bleeding, especially for already
    running user workloads?

        Remove the pod.spec.HostUsers field or disable the feature gate.

  - Diagnostics: What are the useful log messages and their required logging
    levels that could help debug the issue?
    Not required until feature graduated to beta.

>         The error is returned on pod-creation, no need to search for logs.

Reviewer 3: Rather on an attempt to start the pod?

IIUC, Kubelet will get an error from the runtime and will propagate it to the pod, right?

Reviewer: Yes, exactly

Reviewer 3: Can you clarify in the text?

Reviewer: Oh, sorry. Added it now! Here and in the rest of the errors, that this is visible on the pod events.

  - Mitigations: What can be done to stop the bleeding, especially for already
    running user workloads?

        Remove the pod.spec.HostUsers field or disable the feature gate.

  - Diagnostics: What are the useful log messages and their required logging
    levels that could help debug the issue?
    Not required until feature graduated to beta.

        The most likely reason why this can fail is if the kernel doesn't support idmapped mounts.  In this
        case the error reported from the OCI runtime is clear enough, both the runc and crun OCI runtimes were
        instrumented to return a clear error message.

  - Testing: Are there any tests for failure mode? If not, describe why.

        There are extensive unit tests on the function that parses the file content.
        There are no tests for failures to read or write the file, the code-paths just return the errors
        in those cases.

- Error getting the kubelet IDs range configuration
  - Detection: How can it be detected via metrics? Stated another way:
    how can an operator troubleshoot without logging into a master or worker node?

>         In this case the Kubelet will fail to start with a clear error message.

Reviewer 6: Assuming this is through subuids binary don't we fallback to default range?

Author: I think we should fallback to default ranges only if there is no `kubelet` user configured, or there is no `getsubids` binary present.

If the `kubelet` user is defined and we fail to read its configuration, an error should be reported.

>         In this case the Kubelet will fail to start with a clear error message.

Reviewer 7: (nit)
```suggestion
        In this case the kubelet will fail to start with a clear error message.
```

### 64k mappings?

We discussed using shorter or even allowing for longer mappings in the past. The decision is to use
64k mappings (IDs from 0-65535 are mapped/valid in the pod).

The reasons to consider smaller mappings were valid only before idmap mounts was merged into the
kernel. However, idmap mounts is merged for some years now and we require it, making those reasons
void.

The issues without idmap mounts in previous iterations of this KEP, is that the IDs assigned to a
pod had to be unique for every pod in the cluster, easily reaching a limit when the cluster is "big
enough" and the UID space runs out.  However, with idmap mounts the IDs assigned to a pod just needs
to be unique within the node (and with 64k ranges we have 64k pods possible in the node, so not

> really an issue). IOW, by using idmap mounts, we changed the IDs limit to be node-scoped instead of

Reviewer 7: (nit)
```suggestion
really an issue). In other words: by using idmap mounts, we changed the IDs limit to be node-scoped instead of
```

cluster-wide/cluster-scoped.

There are no known use cases for longer mappings that we know of. The 16bit range (0-65535) is what
is assumed by all POSIX tools that we are aware of. If the need arises, longer mapping can be

> considered in a future KEP.

Reviewer 7: _no_ known use cases?

Two examples come to mind:
- running a container tool inside a Pod, where that container tool wants to use a UID range
- running an application inside a Pod where the application uses UIDs above 65535 by default; I've worked with some of these, usually avoiding the 0-65535 range to avoid the risk of conflict with a local system user.

Reviewer: 1. The first example doesn't necessarily need bigger mappings.
1. We didn't hear of any such application that wasn't possible to configure it. The point stands: that is out of scope for this KEP, it can be worked on in a future KEP.

disagree on this, we still need to discuss it and settle on something.  This was
[raised here](https://github.com/kubernetes/enhancements/pull/3065#discussion_r798760382)

> Furthermore, the reasons mentioned by Tim (some nodes having CRIO, some others having containerd,

Reviewer 7: Let's specify which Tim. The Kubernetes project has multiple Tims :wink:
